File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/InstallActivity.java
package de.marmaro.krt.ffupdater;

import android.app.DownloadManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.IntentSender;
import android.content.SharedPreferences;
import android.content.pm.PackageInstaller;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.os.StatFs;
import android.util.Log;
import android.view.MenuItem;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.preference.PreferenceManager;

import org.apache.commons.codec.binary.ApacheCodecHex;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.time.Duration;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import de.marmaro.krt.ffupdater.device.DeviceEnvironment;
import de.marmaro.krt.ffupdater.download.DownloadManagerAdapter;
import de.marmaro.krt.ffupdater.download.DownloadManagerAdapter.StatusProgress;
import de.marmaro.krt.ffupdater.metadata.AvailableMetadata;
import de.marmaro.krt.ffupdater.metadata.AvailableMetadataFetcher;
import de.marmaro.krt.ffupdater.metadata.InstalledMetadataRegister;
import de.marmaro.krt.ffupdater.security.FingerprintValidator;
import de.marmaro.krt.ffupdater.security.FingerprintValidator.FingerprintResult;
import de.marmaro.krt.ffupdater.settings.SettingsHelper;
import de.marmaro.krt.ffupdater.utils.Utils;

import static android.app.DownloadManager.Request.VISIBILITY_VISIBLE;
import static android.app.DownloadManager.STATUS_FAILED;
import static android.app.DownloadManager.STATUS_PAUSED;
import static android.app.DownloadManager.STATUS_PENDING;
import static android.app.DownloadManager.STATUS_RUNNING;
import static android.app.DownloadManager.STATUS_SUCCESSFUL;
import static android.content.pm.PackageInstaller.SessionParams.MODE_FULL_INSTALL;
import static android.os.Environment.DIRECTORY_DOWNLOADS;
import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static de.marmaro.krt.ffupdater.R.id.downloadedFileUrl;

/**
 * Activity for downloading and installing apps on devices with API Level >= 24/Nougat.
 * Reason: If have to use the DownloadManager because this is the easiest way to download the app and access it with
 * the scheme format (for example: content://downloads/all_downloads/20).
 * The DownloadManager is more difficult to use then the default java way, but the DownloadManager offers more features
 * like restarting downloads, showing the current download status etc.
 */
public class InstallActivity extends AppCompatActivity {
    public static final String EXTRA_APP_NAME = "app_name";
    private static final String PACKAGE_INSTALLED_ACTION = "de.marmaro.krt.ffupdater.InstallActivity.SESSION_API_PACKAGE_INSTALLED";
    public static final String LOG_TAG = "InstallActivity";

    private final Map<Integer, String> downloadManagerIdToString = new HashMap<>();
    private DownloadManagerAdapter downloadManager;
    private InstalledMetadataRegister installedMetadataRegister;
    private FingerprintValidator fingerprintValidator;
    private AvailableMetadataFetcher availableMetadataFetcher;

    private App app;
    private AvailableMetadata metadata;
    private long downloadId = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.download_activity);
        AppCompatDelegate.setDefaultNightMode(SettingsHelper.getThemePreference(this, new DeviceEnvironment()));
        registerReceiver(onDownloadComplete, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
        Optional.ofNullable(getSupportActionBar()).ifPresent(actionBar -> actionBar.setDisplayHomeAsUpEnabled(true));

        final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        downloadManager = new DownloadManagerAdapter((DownloadManager) getSystemService(DOWNLOAD_SERVICE));
        installedMetadataRegister = new InstalledMetadataRegister(getPackageManager(), preferences);
        availableMetadataFetcher = new AvailableMetadataFetcher(preferences, new DeviceEnvironment());
        fingerprintValidator = new FingerprintValidator(getPackageManager());
        downloadManagerIdToString.put(STATUS_RUNNING, "running");
        downloadManagerIdToString.put(STATUS_SUCCESSFUL, "success");
        downloadManagerIdToString.put(STATUS_FAILED, "failed");
        downloadManagerIdToString.put(STATUS_PAUSED, "paused");
        downloadManagerIdToString.put(STATUS_PENDING, "pending");

        final Optional<App> optionalApp = Optional.ofNullable(getIntent().getExtras())
                .map(bundle -> bundle.getString(EXTRA_APP_NAME))
                .map(App::valueOf);
        if (!optionalApp.isPresent()) { // if the Activity was not crated from the FFUpdater main menu
            finish();
            return;
        }
        app = optionalApp.get();

        findViewById(R.id.installConfirmationButton).setOnClickListener(v -> install());

        if (isSignatureOfInstalledAppUnknown(app) || isExternalStorageNotAccessible()) {
            return;
        }
        checkFreeSpace();
        fetchAvailableMetadata();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unregisterReceiver(onDownloadComplete);
        availableMetadataFetcher.shutdown();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private boolean isSignatureOfInstalledAppUnknown(App app) {
        if (!installedMetadataRegister.isInstalled(app) || fingerprintValidator.checkInstalledApp(app).isValid()) {
            return false;
        }
        show(R.id.unknownSignatureOfInstalledApp);
        return true;
    }

    private boolean isExternalStorageNotAccessible() {
        final String status = Environment.getExternalStorageState();
        if (Environment.MEDIA_MOUNTED.equals(status)) {
            return false;
        }
        show(R.id.externalStorageNotAccessible);
        setText(R.id.externalStorageNotAccessible_state, status);
        return true;
    }

    private void checkFreeSpace() {
        final File downloadDir = Objects.requireNonNull(getExternalFilesDir(DIRECTORY_DOWNLOADS));
        long freeBytes = new StatFs(downloadDir.getPath()).getFreeBytes();
        if (freeBytes > 104_857_600) {
            return;
        }

        show(R.id.tooLowMemory);
        long freeMBytes = freeBytes / (1024 * 1024);
        setText(R.id.tooLowMemoryDescription, getString(R.string.too_low_memory_description, freeMBytes));
    }

    private void fetchAvailableMetadata() {
        show(R.id.fetchUrl);
        setText(R.id.fetchUrlTextView, getString(
                R.string.fetch_url_for_download,
                app.getDownloadSource(this)));

        final List<App> apps = Collections.singletonList(app);
        final Future<AvailableMetadata> futureMetadata = availableMetadataFetcher.fetchMetadata(apps).get(app);
        new Thread(() -> {
            try {
                this.metadata = futureMetadata.get(30, TimeUnit.SECONDS);
                hide(R.id.fetchUrl);
                show(R.id.fetchedUrlSuccess);
                setText(R.id.fetchedUrlSuccessTextView, getString(R.string.fetched_url_for_download_successfully,
                        app.getDownloadSource(this)));
                downloadApplication();
            } catch (ExecutionException | InterruptedException | TimeoutException e) {
                Log.e(LOG_TAG, "failed to fetch url", e);
                hide(R.id.fetchUrl);
                show(R.id.fetchedUrlFailure);
                setText(R.id.fetchedUrlFailureTextView, getString(R.string.fetched_url_for_download_unsuccessfully,
                        app.getDownloadSource(this)));
                show(R.id.installerFailed);
            }
        }).start();
    }

    private void downloadApplication() {
        show(R.id.downloadingFile);
        setText(R.id.downloadingFileUrl, metadata.getDownloadUrl().toString());

        downloadId = downloadManager.enqueue(this, metadata.getDownloadUrl(), app.getTitle(this), VISIBILITY_VISIBLE);
        new Thread(() -> {
            int previousStatus = -1;
            long waitMs = 500;
            long maxTime = Duration.ofMinutes(5).toMillis();
            for (long i = 0; i < maxTime / waitMs; i++) {
                StatusProgress result = downloadManager.getStatusAndProgress(downloadId);
                if (previousStatus != result.getStatus()) {
                    previousStatus = result.getStatus();
                    setText(R.id.downloadingFileText, getString(R.string.download_application_from_with_status,
                            downloadManagerIdToString.getOrDefault(result.getStatus(), "?")));
                }
                if (result.getStatus() == STATUS_FAILED || result.getStatus() == STATUS_SUCCESSFUL) {
                    return;
                }
                runOnUiThread(() -> ((ProgressBar) findViewById(R.id.downloadingFileProgressBar))
                        .setProgress(result.getProgress()));
                Utils.sleepAndIgnoreInterruptedException(waitMs);
            }
        }).start();
    }

    private final BroadcastReceiver onDownloadComplete = new BroadcastReceiver() {
        public void onReceive(Context context, Intent intent) {
            long id = Objects.requireNonNull(intent.getExtras()).getLong(DownloadManager.EXTRA_DOWNLOAD_ID);
            if (id != downloadId) {
                // received an older message - skip
                return;
            }
            if (downloadManager.getStatusAndProgress(id).getStatus() == STATUS_FAILED) {
                hide(R.id.downloadingFile);
                show(R.id.downloadFileFailed);
                setText(R.id.downloadFileFailedUrl, metadata.getDownloadUrl().toString());
                show(R.id.installerFailed);
                return;
            }
            hide(R.id.downloadingFile);
            show(R.id.downloadedFile);
            setText(downloadedFileUrl, metadata.getDownloadUrl().toString());
            show(R.id.verifyDownloadFingerprint);
            new Thread(() -> {
                File downloadedFile = downloadManager.getFileForDownloadedFile(id);
                final FingerprintResult result = fingerprintValidator.checkApkFile(downloadedFile, app);
                if (result.isValid()) {
                    hide(R.id.verifyDownloadFingerprint);
                    show(R.id.fingerprintDownloadGood);
                    setText(R.id.fingerprintDownloadGoodHash, result.getHexString());
                    show(R.id.installConfirmation);
                } else {
                    hide(R.id.verifyDownloadFingerprint);
                    show(R.id.fingerprintDownloadBad);
                    setText(R.id.fingerprintDownloadBadHashActual, result.getHexString());
                    setText(R.id.fingerprintDownloadBadHashExpected, ApacheCodecHex.encodeHexString(app.getSignatureHash()));
                    show(R.id.installerFailed);
                }
            }).start();
        }
    };

    /**
     * See example: https://android.googlesource.com/platform/development/+/master/samples/ApiDemos/src/com/example/android/apis/content/InstallApkSessionApi.java
     */
    private void install() {
        show(R.id.installingApplication);
        PackageInstaller installer = getPackageManager().getPackageInstaller();
        PackageInstaller.SessionParams params = new PackageInstaller.SessionParams(MODE_FULL_INSTALL);
        try (PackageInstaller.Session session = installer.openSession(installer.createSession(params))) {
            int lengthInBytes = downloadManager.getTotalDownloadSize(downloadId);
            Uri download = downloadManager.getUriForDownloadedFile(downloadId);

            try (OutputStream packageInSession = session.openWrite("package", 0, lengthInBytes);
                 InputStream apk = getContentResolver().openInputStream(download)) {
                Objects.requireNonNull(apk);
                byte[] buffer = new byte[16384];
                int n;
                while ((n = apk.read(buffer)) >= 0) {
                    packageInSession.write(buffer, 0, n);
                }
            }

            Context context = InstallActivity.this;
            Intent intent = new Intent(context, InstallActivity.class);
            intent.setAction(PACKAGE_INSTALLED_ACTION);
            PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);
            IntentSender intentSender = pendingIntent.getIntentSender();
            session.commit(intentSender);
        } catch (IOException e) {
            Log.e(LOG_TAG, "failed to install APK", e);
        }
    }

    /**
     * This method will be called when the app installation is completed.
     *
     * @param intent intent
     */
    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (PACKAGE_INSTALLED_ACTION.equals(intent.getAction())) {
            Bundle extras = Objects.requireNonNull(intent.getExtras());
            int status = extras.getInt(PackageInstaller.EXTRA_STATUS);
            if (status == PackageInstaller.STATUS_PENDING_USER_ACTION) {
                // This test app isn't privileged, so the user has to confirm the install.
                startActivity((Intent) extras.get(Intent.EXTRA_INTENT));
                return;
            }
            hide(R.id.installingApplication);
            hide(R.id.installConfirmation);
            if (status == PackageInstaller.STATUS_SUCCESS) {
                show(R.id.installerSuccess);
                actionVerifyInstalledAppSignature();
                installedMetadataRegister.saveReleaseId(app, metadata.getReleaseId());
            } else {
                show(R.id.installerFailed);
                String message = extras.getString(PackageInstaller.EXTRA_STATUS_MESSAGE);
                String text = String.format(Locale.getDefault(), "(%d) %s", status, message);
                setText(R.id.installerFailedReason, text);
            }
            downloadManager.remove(downloadId);
        }
    }

    private void actionVerifyInstalledAppSignature() {
        show(R.id.verifyInstalledFingerprint);
        new Thread(() -> {
            final FingerprintResult fingerprintResult = fingerprintValidator.checkInstalledApp(app);
            hide(R.id.verifyInstalledFingerprint);
            if (fingerprintResult.isValid()) {
                show(R.id.fingerprintInstalledGood);
                setText(R.id.fingerprintInstalledGoodHash, fingerprintResult.getHexString());
            } else {
                show(R.id.fingerprintInstalledBad);
                setText(R.id.fingerprintInstalledBadHashActual, fingerprintResult.getHexString());
                setText(R.id.fingerprintInstalledBadHashExpected, ApacheCodecHex.encodeHexString(app.getSignatureHash()));
            }
        }).start();
    }

    private void show(int viewId) {
        runOnUiThread(() -> findViewById(viewId).setVisibility(VISIBLE));
    }

    private void hide(int viewId) {
        runOnUiThread(() -> findViewById(viewId).setVisibility(GONE));
    }

    private void setText(int textId, String text) {
        runOnUiThread(() -> ((TextView) findViewById(textId)).setText(text));
    }
}

File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/InstallActivity.kt
package de.marmaro.krt.ffupdater

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.os.FileUriExposedException
import android.os.StrictMode
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import android.widget.Toast.LENGTH_LONG
import androidx.annotation.MainThread
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.viewModelScope
import de.marmaro.krt.ffupdater.R.string.*
import de.marmaro.krt.ffupdater.app.App
import de.marmaro.krt.ffupdater.app.UpdateCheckResult
import de.marmaro.krt.ffupdater.app.impl.exceptions.GithubRateLimitExceededException
import de.marmaro.krt.ffupdater.app.impl.exceptions.NetworkException
import de.marmaro.krt.ffupdater.background.BackgroundNotificationBuilder
import de.marmaro.krt.ffupdater.crash.CrashListener
import de.marmaro.krt.ffupdater.device.DeviceSdkTester
import de.marmaro.krt.ffupdater.download.AppCache
import de.marmaro.krt.ffupdater.download.FileDownloader
import de.marmaro.krt.ffupdater.download.NetworkUtil.isNetworkMetered
import de.marmaro.krt.ffupdater.download.StorageUtil
import de.marmaro.krt.ffupdater.installer.ForegroundAppInstaller
import de.marmaro.krt.ffupdater.settings.ForegroundSettingsHelper
import de.marmaro.krt.ffupdater.settings.InstallerSettingsHelper
import de.marmaro.krt.ffupdater.settings.InstallerSettingsHelper.Installer.SESSION_INSTALLER
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext


/**
 * Activity for downloading and installing apps on devices with API Level >= 24/Nougat.
 * Reason: If have to use the DownloadManager because this is the easiest way to download the app and access it with
 * the scheme format (for example: content://downloads/all_downloads/20).
 * The DownloadManager is more difficult to use then the default java way, but the DownloadManager offers more features
 * like restarting downloads, showing the current download status etc.
 */
class InstallActivity : AppCompatActivity() {
    private lateinit var viewModel: InstallActivityViewModel
    private lateinit var appInstaller: ForegroundAppInstaller
    private lateinit var appCache: AppCache
    private lateinit var foregroundSettings: ForegroundSettingsHelper
    private lateinit var installerSettingsHelper: InstallerSettingsHelper

    // persistent data across orientation changes
    class InstallActivityViewModel : ViewModel() {
        var app: App? = null
        var fileDownloader: FileDownloader? = null
        var updateCheckResult: UpdateCheckResult? = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.install_activity)
        if (CrashListener.openCrashReporterForUncaughtExceptions(this)) {
            finish()
            return
        }
        AppCompatDelegate.setDefaultNightMode(ForegroundSettingsHelper(this).themePreference)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        viewModel = ViewModelProvider(this).get(InstallActivityViewModel::class.java)
        if (viewModel.app == null) {
            viewModel.app = intent.extras?.getString(EXTRA_APP_NAME)
                ?.let { App.valueOf(it) }
                ?: run {
                    // InstallActivity was unintentionally started again after finishing the download
                    finish()
                    return
                }
        }

        foregroundSettings = ForegroundSettingsHelper(this)
        installerSettingsHelper = InstallerSettingsHelper(this)
        appCache = AppCache(viewModel.app!!)
        appInstaller = ForegroundAppInstaller.create(this, viewModel.app!!, appCache.getFile(this))
        lifecycle.addObserver(appInstaller)

        findViewById<Button>(R.id.install_activity__delete_cache_button).setOnClickListener {
            appCache.delete(this)
            hide(R.id.install_activity__delete_cache)
        }
        findViewById<Button>(R.id.install_activity__open_cache_folder_button).setOnClickListener {
            tryOpenDownloadFolderInFileManager()
        }

        // hide existing background notification for this app
        BackgroundNotificationBuilder.hideUpdateIsAvailable(this, viewModel.app!!)
        BackgroundNotificationBuilder.hideInstallationSuccess(this, viewModel.app!!)
        BackgroundNotificationBuilder.hideInstallationError(this, viewModel.app!!)

        lifecycleScope.launch(Dispatchers.Main) {
            startInstallationProcess()
        }
    }

    private fun tryOpenDownloadFolderInFileManager() {
        val intent = Intent(Intent.ACTION_VIEW)
        val parentFolder = appCache.getFile(this).parentFile ?: return
        val uri = Uri.parse("file://${parentFolder.absolutePath}/")
        intent.setDataAndType(uri, "resource/folder")
        val chooser = Intent.createChooser(intent, getString(install_activity__open_folder))

        if (DeviceSdkTester.supportsAndroidNougat()) {
            StrictMode.setVmPolicy(StrictMode.VmPolicy.LAX)
            try {
                startActivity(chooser)
            } catch (e: FileUriExposedException) {
                Toast.makeText(this, install_activity__file_uri_exposed_toast, LENGTH_LONG)
                    .show()
            }
        } else {
            startActivity(chooser)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun show(viewId: Int) {
        findViewById<View>(viewId).visibility = View.VISIBLE
    }

    private fun hide(viewId: Int) {
        findViewById<View>(viewId).visibility = View.GONE
    }

    private fun setText(textId: Int, text: String) {
        findViewById<TextView>(textId).text = text
    }

    private suspend fun startInstallationProcess() {
        checkIfStorageIsMounted()
    }

    private suspend fun checkIfStorageIsMounted() {
        if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED) {
            checkIfEnoughStorageAvailable()
            return
        }
        failureExternalStorageNotAccessible()
    }

    private suspend fun checkIfEnoughStorageAvailable() {
        if (StorageUtil.isEnoughStorageAvailable(this)) {
            fetchDownloadInformation()
            return
        }
        show(R.id.tooLowMemory)
        val mbs = StorageUtil.getFreeStorageInMebibytes(this)
        val message = getString(install_activity__too_low_memory_description, mbs)
        setText(R.id.tooLowMemoryDescription, message)
        fetchDownloadInformation()
    }

    private suspend fun fetchDownloadInformation() {
        show(R.id.fetchUrl)
        val downloadSource = getString(viewModel.app!!.detail.displayDownloadSource)
        val runningText = getString(install_activity__fetch_url_for_download, downloadSource)
        setText(R.id.fetchUrlTextView, runningText)

        // check if network type requirements are met
        if (!foregroundSettings.isUpdateCheckOnMeteredAllowed && isNetworkMetered(this)) {
            failureShowFetchUrlException(getString(main_activity__no_unmetered_network))
            return
        }

        val updateCheckResult = try {
            viewModel.app!!.detail.updateCheckAsync(this).await()
        } catch (e: GithubRateLimitExceededException) {
            failureShowFetchUrlException(getString(install_activity__github_rate_limit_exceeded), e)
            return
        } catch (e: NetworkException) {
            failureShowFetchUrlException(getString(install_activity__temporary_network_issue), e)
            return
        }

        viewModel.updateCheckResult = updateCheckResult
        hide(R.id.fetchUrl)
        show(R.id.fetchedUrlSuccess)
        val finishedText = getString(install_activity__fetched_url_for_download_successfully, downloadSource)
        setText(R.id.fetchedUrlSuccessTextView, finishedText)
        if (appCache.isAvailable(this, updateCheckResult.availableResult)) {
            show(R.id.useCachedDownloadedApk)
            setText(R.id.useCachedDownloadedApk__path, appCache.getFile(this).absolutePath)
            installApp()
            return
        }

        if (viewModel.fileDownloader?.currentDownload != null) {
            reuseCurrentDownload()
            return
        }
        startDownload()
    }

    @MainThread
    private suspend fun startDownload() {
        if (!foregroundSettings.isDownloadOnMeteredAllowed && isNetworkMetered(this)) {
            failureShowFetchUrlException(getString(main_activity__no_unmetered_network))
            return
        }

        val updateCheckResult = requireNotNull(viewModel.updateCheckResult)
        show(R.id.downloadingFile)
        setText(R.id.downloadingFileUrl, updateCheckResult.downloadUrl)

        val fileDownloader = FileDownloader()
        viewModel.fileDownloader = fileDownloader
        fileDownloader.onProgress = { percentage, mb ->
            runOnUiThread {
                val status = if (percentage != null) {
                    findViewById<ProgressBar>(R.id.downloadingFileProgressBar).progress = percentage
                    "${percentage}%"
                } else {
                    "${mb}MB"
                }
                val display = getString(install_activity__download_app_with_status, status)
                setText(R.id.downloadingFileText, display)
            }
        }

        val display = getString(install_activity__download_app_with_status, "")
        setText(R.id.downloadingFileText, display)

        val url = updateCheckResult.availableResult.downloadUrl
        appCache.delete(this)
        val file = appCache.getFile(this)

        // this coroutine should survive a screen rotation and should live as long as the view model
        try {
            withContext(viewModel.viewModelScope.coroutineContext) {
                fileDownloader.downloadFileAsync(url, file).await()
            }
            hide(R.id.downloadingFile)
            show(R.id.downloadedFile)
            setText(R.id.downloadedFileUrl, updateCheckResult.downloadUrl)
            installApp()
        } catch (e: NetworkException) {
            hide(R.id.downloadingFile)
            show(R.id.downloadedFile)
            setText(R.id.downloadedFileUrl, updateCheckResult.downloadUrl)
            failureDownloadUnsuccessful(e)
        }
    }

    @MainThread
    private suspend fun reuseCurrentDownload() {
        val updateCheckResult = requireNotNull(viewModel.updateCheckResult)
        show(R.id.downloadingFile)
        setText(R.id.downloadingFileUrl, updateCheckResult.downloadUrl)
        val fileDownloader = requireNotNull(viewModel.fileDownloader)
        fileDownloader.onProgress = { percentage, mb ->
            runOnUiThread {
                val status = if (percentage != null) {
                    findViewById<ProgressBar>(R.id.downloadingFileProgressBar).progress = percentage
                    "${percentage}%"
                } else {
                    "${mb}MB"
                }
                val display = getString(install_activity__download_app_with_status, status)
                setText(R.id.downloadingFileText, display)
            }
        }

        try {
            fileDownloader.currentDownload?.await()
            installApp()
        } catch (e: NetworkException) {
            failureDownloadUnsuccessful(e)
        }
    }

    @MainThread
    private suspend fun installApp() {
        show(R.id.installingApplication)
        val result = appInstaller.installAsync(this).await()

        // hide existing background notification for this app
        BackgroundNotificationBuilder.hideUpdateIsAvailable(this, viewModel.app!!)
        BackgroundNotificationBuilder.hideInstallationSuccess(this, viewModel.app!!)
        BackgroundNotificationBuilder.hideInstallationError(this, viewModel.app!!)

        if (result.success) {
            hide(R.id.installingApplication)
            show(R.id.installerSuccess)
            show(R.id.fingerprintInstalledGood)
            setText(R.id.fingerprintInstalledGoodHash, result.certificateHash ?: "/")

            val available = requireNotNull(viewModel.updateCheckResult).availableResult
            viewModel.app!!.detail.appIsInstalled(this, available)

            if (foregroundSettings.isDeleteUpdateIfInstallSuccessful) {
                appCache.delete(this)
            } else {
                show(R.id.install_activity__delete_cache)
            }
            return
        }

        hide(R.id.installingApplication)
        show(R.id.installerFailed)
        if (installerSettingsHelper.getInstaller() == SESSION_INSTALLER) {
            show(R.id.install_activity__different_installer_info)
        }

        show(R.id.install_activity__open_cache_folder)
        val cacheFolder = appCache.getFile(this).parentFile?.absolutePath ?: ""
        setText(R.id.install_activity__cache_folder_path, cacheFolder)
        setText(R.id.installerFailedReason, result.errorMessage ?: "/")

        if (foregroundSettings.isDeleteUpdateIfInstallFailed) {
            appCache.delete(this)
        } else {
            show(R.id.install_activity__delete_cache)
        }

        if (result.errorException != null) {
            findViewById<TextView>(R.id.installerFailedReason).setOnClickListener {
                val description = getString(crash_report___explain_text__install_activity_install_file)
                val intent = CrashReportActivity.createIntent(this, result.errorException, description)
                startActivity(intent)
            }
        }
    }

    @MainThread
    private fun failureExternalStorageNotAccessible() {
        show(R.id.externalStorageNotAccessible)
        setText(R.id.externalStorageNotAccessible_state, Environment.getExternalStorageState())
    }

    @MainThread
    private fun failureDownloadUnsuccessful(exception: Exception) {
        val updateCheckResult = requireNotNull(viewModel.updateCheckResult)
        hide(R.id.downloadingFile)
        show(R.id.downloadFileFailed)
        setText(R.id.downloadFileFailedUrl, updateCheckResult.downloadUrl)

        findViewById<TextView>(R.id.downloadFileFailedShowException).setOnClickListener {
            val description = getString(crash_report___explain_text__install_activity_download_file)
            val intent = CrashReportActivity.createIntent(this, exception, description)
            startActivity(intent)
        }
        show(R.id.installerFailed)
        appCache.delete(this)
    }

    @MainThread
    private fun failureShowFetchUrlException(message: String, exception: Exception? = null) {
        hide(R.id.fetchUrl)
        show(R.id.install_activity__exception)
        setText(R.id.install_activity__exception__text, message)
        if (exception == null) {
            hide(R.id.install_activity__exception__show_button)
            return
        }

        findViewById<TextView>(R.id.install_activity__exception__show_button).setOnClickListener {
            val description = getString(crash_report__explain_text__install_activity_fetching_url)
            val intent = CrashReportActivity.createIntent(this, exception, description)
            startActivity(intent)
        }
    }

    companion object {
        const val EXTRA_APP_NAME = "app_name"
        fun createIntent(context: Context, app: App): Intent {
            val intent = Intent(context, InstallActivity::class.java)
            // intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            intent.putExtra(EXTRA_APP_NAME, app.name)
            return intent
        }
    }
}